File: packages/conab-client/src/resolve-url.test.ts
import { describe, expect, it } from 'vitest';

import { resolveDownloadUrl } from './resolve-url.js';

describe('resolveDownloadUrl', () => {
  it('resolves estimativa-graos from catalog', () => {
    const resolved = resolveDownloadUrl('conab.estimativa-graos');
    expect(resolved.url).toBe(
      'https://portaldeinformacoes.conab.gov.br/downloads/arquivos/LevantamentoGraos.txt',
    );
    expect(resolved.portalPage).toContain('download-arquivos.html');
  });

  it('throws for unknown dataset', () => {
    expect(() => resolveDownloadUrl('unknown.dataset')).toThrow('No CONAB download mapping');
  });
});

File: internal/conab/resolve_url_test.go
package conab_test

import (
	"testing"

	"github.com/open-data-brazil/open-data-agro/internal/catalog"
	"github.com/open-data-brazil/open-data-agro/internal/conab"
)

func TestResolveURLValidatesPortalHost(t *testing.T) {
	t.Parallel()

	entry := catalog.RegistryEntry{
		DatasetID: catalog.MustParseDatasetID("conab.estimativa-graos"),
		SourceURL: "https://portaldeinformacoes.conab.gov.br/downloads/arquivos/LevantamentoGraos.txt",
	}
	url, err := conab.ResolveURL(entry)
	if err != nil {
		t.Fatalf("ResolveURL: %v", err)
	}
	if url != entry.SourceURL {
		t.Fatalf("got %q want %q", url, entry.SourceURL)
	}
}

func TestResolveURLRejectsExternalHost(t *testing.T) {
	t.Parallel()

	entry := catalog.RegistryEntry{
		DatasetID: catalog.MustParseDatasetID("conab.estimativa-graos"),
		SourceURL: "https://example.com/file.txt",
	}
	if _, err := conab.ResolveURL(entry); err == nil {
		t.Fatal("expected error for external host")
	}
}
